aea/agent_loop.py

```diff
@@ -0,0 +1,406 @@
+        self._exception_callback = exception_callback
+
+    def _callback(self) -> None:
+        """Call on each schduled call."""
+        self._schedule_call()
+        try:
+            self._periodic_callable()
+        except Exception as exception:
+            if not self._exception_callback:
+                raise
+            self._exception_callback(self._periodic_callable, exception)
+
+    def _schedule_call(self) -> None:
+        """Set schedule for call."""
+        if self._period is None:
+            return
+
+        if self._timerhandle is None:
+            ts = time.mktime(self._start_at.timetuple())
+            delay = max(0, ts - time.time())
+            self._timerhandle = self._loop.call_later(delay, self._callback)
+        else:
+            self._timerhandle = self._loop.call_later(self._period, self._callback)
+
+    def start(self) -> None:
+        """Activate periodi calls."""
+        if self._timerhandle:
+            return
+        self._schedule_call()
+
+    def stop(self) -> None:
+        """Remove from schedule."""
+        if not self._timerhandle:
+            return
+
+        self._timerhandle.cancel()
+        self._timerhandle = None
+
+
+class BaseAgentLoop(ABC):
+    """Base abstract  agent loop class."""
+
+    def __init__(self, agent: "Agent") -> None:
+        """Init loop.
+
+        :params agent: Agent or AEA to run.
+        """
+        self._agent = agent
+
+    @abstractmethod
+    def start(self) -> None:
+        """Start agent loop."""
+        raise NotImplementedError
+
+    @abstractmethod
+    def stop(self) -> None:
+        """Stop agent loop."""
+        raise NotImplementedError
+
+
+class AgentLoopException(AEAException):
+    """Exception for agent loop runtime errors."""
+
+
+class AsyncAgentLoop(BaseAgentLoop):
+    """Asyncio based agent loop suitable only for AEA."""
+
+    NEW_BEHAVIOURS_PROCESS_SLEEP = 1  # check new behaviours registered every second.
+
+    def __init__(self, agent: "AEA", loop: AbstractEventLoop = None):
+        """
+        Init agent loop.
+
+        :param agent: AEA instance
+        :param loop: asyncio loop to use. optional
+        """
+        super().__init__(agent)
+        self._agent: "AEA" = self._agent
+
+        try:
+            self._loop = loop or asyncio.get_event_loop()
+            assert not self._loop.is_closed()
+            assert not self._loop.is_running()
+        except (RuntimeError, AssertionError):
+            self._loop = asyncio.new_event_loop()
+            asyncio.set_event_loop(self._loop)
+
+        self._behaviours_registry: Dict[Behaviour, PeriodicCaller] = {}
+        self._state: AsyncState = AsyncState()
+        self._exceptions: List[Exception] = []
+
+    def start(self):
+        """Start agent loop."""
+        self._state.state = "started"
+        self._loop.run_until_complete(self._run())
+
+    def stop(self):
+        """Stop agent loop."""
+        self._state.state = "stopping"
```

Comment:
we could use an `enum.Enum` for the possible state values?

Reply:
I'm aware it requires non-trivial extra work, so take it as a suggestion, and if we have no time then create an issue if you consider it worthwhile.

Reply:
it does not require extra work, but it's not assumed somebody will set state manually.
so it's internal states and I'm not sure enum is really required here.

Reply:
done